---
title: "Inquiline Genomics: Analyse CAFE modelling"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 2
    collapsed: true
    number_sections: true
    df_print: paged
  pdf_document: default
---
<STYLE TYPE="text/css">
<!--
  td{
    font-family: Arial; 
    font-size: 8pt;
    padding:0px;
    cellpadding="0";
    cellspacing="0"
  }
  th {
    font-family: Arial; 
    font-size: 8pt;
    height: 20px;
    font-weight: bold;
    text-align: right;
    background-color: #ccccff;
  }
  table { 
    border-spacing: 0px;
    border-collapse: collapse;
  }
--->
</STYLE>


```{r message=FALSE, warning=FALSE, include=FALSE}
source("~/sciebo/librarySchrader.R")
library(dplyr)
library(plotrix)
```

```{r message=FALSE, warning=FALSE, include=FALSE}
nodeLabels<-read.csv("/Users/lukas/sciebo/inquilineGenomics18/geneFamilies/CAFE/results/Acro.Atta/nodeLabels.tsv",sep="\t",header=F)
colnames(nodeLabels)<-c("tree","label")
```

# 2-parameter Models
## Load all 2-parameter models
Load 2-parameter models from RUN5 (contains 8 independent runs for each 2-parameter model). These have been run iteratively to estimate individual rates for each branch in the analyzed phylogeny. Clustering in subsequent steps is based on these 2-parameter model estimates.

The table provides an overview of all eight repetitions of 2-parameter models for each branch.

```{r echo=FALSE, message=FALSE, warning=FALSE}
###########################################################################
#K means clustering
###########################################################################
# 8 runs per model
ratesRUN5<-read.csv("/Users/lukas/sciebo/inquilineGenomics18/geneFamilies/CAFE/results/Acro.Atta/RUN5/treeRates.RUN5.tsv",sep="\t",skip=1,header=F)

colnamesRates<-c("tree","lambda1","lambda2","score","mu1","mu2","scoreTmp")
colnames(ratesRUN5)<-colnamesRates

rate5<-merge(ratesRUN5,nodeLabels,all.x=T,all.y=T,by.x="tree",by.y="tree")
rate5
```

## Select best fitting 2-parameter models 
The figure shows the distribution of estimates for each branch. The red dot shows the lambda and mu estimates for the best-scoring model.
The data frame shown provides an overview of the best-scoring 2-parameter models for each branch.
```{r echo=FALSE}
#split into different branches
s5<-split(rate5, rate5$label)
##number of runs per model
#data.frame(summary(rate5$label))

#calculate range of lambda and mu estimates for 2-parameter models
rangesL<-data.frame(t(sapply((lapply(s5,"[", , 3)),range)))
rangesM<-data.frame(t(sapply((lapply(s5,"[", , 6)),range)))
ranges<-merge(rangesL,rangesM,by="row.names",all.x=TRUE)
colnames(ranges)<-c("label","minLambda","maxLambda","minMu","maxMu")

#calculate best scoring estimate of lambda and mu for 2-parameter models
lowestScore<-NA
for (i in 1:length(s5)){
  lowestScore<-rbind(lowestScore,s5[[i]][which.min(s5[[i]][,4]),])
  
}
s5summary<-merge(lowestScore,ranges,by="label",all.x=F,,all.y=TRUE)
plot(x=lowestScore$lambda2,y=lowestScore$mu2,col=rgb(1,0,0,.5),pch=19,cex=1,xlab="lambda",ylab="mu",ylim=c(0,0.009),xlim=c(0,0.0025))
for (i in 1:nrow(s5summary)){
segments(x0 = s5summary$minLambda[i],x1=s5summary$maxLambda[i],y0=s5summary$mu2[i],y1=s5summary$mu2[i],lwd=.5)
segments(y0 = s5summary$minMu[i],y1=s5summary$maxMu[i],x0=s5summary$lambda2[i],x1=s5summary$lambda2[i],lwd=.5)
}
l<-s5summary$label
levels(l)[levels(l)=="ACRO"]<-"ACROMYRMEX"
text(s5summary$lambda2,s5summary$mu2+.0003,l,cex=.4)
s5summary

#dev.print(pdf,"/Users/lukas/sciebo/inquilineGenomics18/geneFamilies/CAFE/results/Acro.Atta/RUN5.all8runs.pdf")
#write.table(lowestScore,"/Users/lukas/sciebo/inquilineGenomics18/geneFamilies/CAFE/results/Acro.Atta/RUN5.all8runs.bestFits.tsv",sep="\t",quote=F,row.names=F)
```

## Clustering of 2-parameter models
### K-means clustering (n=1 to n=8) for 2-parameter models.
```{r echo=FALSE}
# choose best estimates of the runs as basis for analysis
rate<-s5summary
#as.matrix(rate[2:nrow(rate),c(4,7)])
nmodels<-nrow(rate)
kclusters<-list()
for (i in 1:8){
  #kclusters[[i]]<-kmeans(as.matrix(rate[1:nmodels,c(4,7)]),i, algorithm = "MacQueen", nstart = 1000)
  #kclusters[[i]]<-kmeans(as.matrix(rate[1:nmodels,c(4,7)]),i, algorithm = "Hartigan-Wong", nstart = 1000)
  #kclusters[[i]]<-kmeans(as.matrix(rate[1:nmodels,c(4,7)]),i, algorithm = "Lloyd", nstart = 1000)
  kclusters[[i]]<-kmeans(as.matrix(rate[1:nmodels,c(4,7)]),i, nstart = 1000)
}
#kclusters
rate$k1<-as.factor(c(kclusters[[1]]$cluster))
rate$k2<-as.factor(c(kclusters[[2]]$cluster))
rate$k3<-as.factor(c(kclusters[[3]]$cluster))
rate$k4<-as.factor(c(kclusters[[4]]$cluster))
rate$k5<-as.factor(c(kclusters[[5]]$cluster))
rate$k6<-as.factor(c(kclusters[[6]]$cluster))
rate$k7<-as.factor(c(kclusters[[7]]$cluster))
rate$k8<-as.factor(c(kclusters[[8]]$cluster))
#rate[,c(1:2,13:20)]

```


### Kmeans Clustering results
The plot shows all different clusterings based on kmeans clustering. Dots show position according to 2-parameter model estimates.
```{r echo=FALSE, fig.height=5, fig.width=10}
par(mfrow=c(2,4))
par(mai=c(.6,.5,0.1,0.1))

for (i in 1:8){
  modelLambda<-kclusters[[i]]$centers[,1]
  modelMu<-kclusters[[i]]$centers[,2]
  
  plot(rate$lambda2,rate$mu2,col=rgb(.5,.5,.5,.2),pch=19,xlab=paste("lambda; Clusters=",i,sep=""),ylab="mu",cex=4,lwd=2,ylim=c(0,0.009),xlim=c(0,0.003))
  points(rate$lambda2,rate$mu2,col=rate[,12+i],pch=1,xlab=paste("lambda; Clusters=",i,sep=""),ylab="mu",cex=4,lwd=2)
  if(i==1){
  text(rate$label,x=rate$lambda2,y=rate$mu2,cex=.8)
  }
  text(rate[,12+i],x=rate$lambda2,y=rate$mu2,cex=1,col=rgb(.3,.3,.3,1),font=2)
}

write.table(rate,"/Users/lukas/sciebo/inquilineGenomics18/geneFamilies/CAFE/results/Acro.Atta/clustered.rates.CAFE.RUN5.bestFits.tsv",quote=F, row.names=F,sep="\t")
dev.print(pdf,"/Users/lukas/sciebo/inquilineGenomics18/geneFamilies/CAFE/results/Acro.Atta/RUN5.Kmeans.clustering.pdf",height=5,width=10)
#(((2,(3,6)4)5,(1,4)5)5,5)
#(((AHEY,(ACHA,PARG)4)5,(AECH,AINS)5)5,ACOL)



```


### pamk unsupervised clustering/hierachical clustering
pamk suggests `pamkResults$nc` clusters. An analysis by hierachical clustering based on Euclidean distances suggests that origins of social parasitism clusters together.  
```{r echo=FALSE, message=FALSE, warning=FALSE}
par(mai=c(.6,.5,0.1,0.1))
# alternative clustering approaches
library(fpc)
mydata<-as.matrix(rate[1:nmodels,c(4,7)])
row.names(mydata)<-rate$label[1:nmodels]
pamkResults<-pamk(mydata,3:9)

d <- dist(mydata, method = "euclidean") # distance matrix
fit <- hclust(d, method="ward.D") 
plot(fit,main="",xlab="")
dev.print(pdf,"/Users/lukas/sciebo/inquilineGenomics18/geneFamilies/CAFE/results/Acro.Atta/RUN5.Kmeans.hclust.pdf",width=4,height=7)
```

# Model selection: higher-parameter models
We ran different clusters in CAFE to determine the best fitting model. A model with 6 clusters produced the best likelihood score. We continued to use this clustering for the final analysis of gene family evolution.

The plot shows these clusters. Numbered dots show position of each branch according to 2-parameter models. Coloring and ellipses are indicative of cluster membership (i.e. 6 colors for 6 clusters). Small red dots show estimates for the six different clusters. x= lambda estimates of gene gains, y= mu estimates of gene loss.
```{r echo=FALSE, fig.height=6, fig.width=4}
runs<-read.csv("/Users/lukas/sciebo/inquilineGenomics18/geneFamilies/CAFE/results/Acro.Atta/allProperRuns.tsv",sep="\t")
runs<-runs[order(runs$likelihood),]
#head(runs,1)

xl<-c(0,0.0025)
yl<-c(0,0.0080)
plot(rate$lambda2,rate$mu2,col=rgb(.5,.5,.5,.2),pch=19,xlab=paste("lambda; Clusters=",6,sep=""),ylab="mu",cex=2,lwd=2,ylim=yl,xlim=xl,xaxt="n",yaxt="n")
axis(1,at=xl)
axis(2,at=yl)
#text(rate$label,x=rate$lambda2,y=rate$mu2,cex=.8)

#runs[1,]


#(((2,(3,6)4)5,(1,4)5)5,5)
segments(x0 = rate$lambda2[rate$label=="AECH"],x1=runs$l1[1],y0=rate$mu2[rate$label=="AECH"],y1=runs$m1[1],lty=1)
segments(x0 = rate$lambda2[rate$label=="AHEY"],x1=runs$l2[1],y0=rate$mu2[rate$label=="AHEY"],y1=runs$m2[1],lty=1)
segments(x0 = rate$lambda2[rate$label=="ACHA"],x1=runs$l3[1],y0=rate$mu2[rate$label=="ACHA"],y1=runs$m3[1],lty=1)
segments(x0 = mean(c(rate$lambda2[rate$label=="ACHA/PARG"],rate$lambda2[rate$label=="AINS"])),x1=runs$l4[1],y0=mean(c(rate$mu2[rate$label=="ACHA/PARG"],rate$mu2[rate$label=="AINS"])),y1=runs$m4[1],lty=1)
segments(x0 = mean(c(rate$lambda2[rate$label=="AHEY/ACHA/PARG"],rate$lambda2[rate$label=="AECH/AINS"],rate$lambda2[rate$label=="ACRO"],rate$lambda2[rate$label=="AACOL"])),x1=runs$l5[1],y0=mean(c(rate$mu2[rate$label=="AHEY/ACHA/PARG"],rate$mu2[rate$label=="AECH/AINS"],rate$mu2[rate$label=="ACRO"],rate$mu2[rate$label=="ACOL"])),y1=runs$m5[1],lty=1)
segments(x0 = rate$lambda2[rate$label=="PARG"],x1=runs$l6[1],y0=rate$mu2[rate$label=="PARG"],y1=runs$m6[1],lty=1)

pointcex<-2.3
pointlwd<-3
textcex<-1.1
cluscol<-goodCols
cluscol[3]<-"gray60"
coloring<-cluscol[c(1,2,3,3,4,3,6,3,2,5)]


#(((2,(3,6)4)5,(1,4)5)5,5)
ellipseCol<-rgb(.95,.95,.95,1)
draw.ellipse(runs$l5[1]+0.000055,runs$m5[1]+0.0001,a=.00045,b=.00058,col=ellipseCol,angle=-5,border=cluscol[3],lwd=3)
draw.ellipse(runs$l4[1]+0.000055,runs$m4[1]+0.00092,a=.00037,b=.00075,col=ellipseCol,angle=-30,border=cluscol[2],lwd=3)

for (i in 1:length(rate$lambda2)){
  points(rate$lambda2[i],rate$mu2[i],col=coloring[i],pch=19,cex=pointcex+1,lwd=pointlwd)
  points(rate$lambda2[i],rate$mu2[i],col="white",pch=19,cex=pointcex,lwd=2)
  #(((2,(3,6)4)5,(1,4)5)5,5)
  text(i,x=rate$lambda2[i],y=rate$mu2[i],cex=textcex,col=1,font=1)
}
points(runs[1,4:9],runs[1,15:20],cex=.5,col="red",pch=19)
points(runs[1,4:9],runs[1,15:20],col="black",pch=1,cex=.5)


text(runs$l1[1],runs$m1[1],paste("x = ",sprintf("%.5f",runs$l1[1]),"\ny = ",sprintf("%.5f",runs$m1[1]),sep=""),cex=.5)
text(runs$l2[1],runs$m2[1],paste("x = ",sprintf("%.5f",runs$l2[1]),"\ny = ",sprintf("%.5f",runs$m2[1]),sep=""),cex=.5)
text(runs$l3[1],runs$m3[1],paste("x = ",sprintf("%.5f",runs$l3[1]),"\ny = ",sprintf("%.5f",runs$m3[1]),sep=""),cex=.5)
text(runs$l4[1],runs$m4[1],paste("x = ",sprintf("%.5f",runs$l4[1]),"\ny = ",sprintf("%.5f",runs$m4[1]),sep=""),cex=.5)
text(runs$l5[1],runs$m5[1],paste("x = ",sprintf("%.5f",runs$l5[1]),"\ny = ",sprintf("%.5f",runs$m5[1]),sep=""),cex=.5)
text(runs$l6[1],runs$m6[1],paste("x = ",sprintf("%.5f",runs$l6[1]),"\ny = ",sprintf("%.5f",runs$m6[1]),sep=""),cex=.5)

dev.print(pdf,"/Users/lukas/sciebo/inquilineGenomics18/geneFamilies/CAFE/results/Acro.Atta/finalClustering.pdf",width = 4, height = 5.5)
```



