---
title: "Inquiline Genomics: GO enrichment analysis of RELAX results"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 2
    collapsed: true
    number_sections: true
    df_print: paged
  pdf_document: default
---
<STYLE TYPE="text/css">
<!--
  td{
    font-family: Arial; 
    font-size: 8pt;
    padding:0px;
    cellpadding="0";
    cellspacing="0"
  }
  th {
    font-family: Arial; 
    font-size: 8pt;
    height: 20px;
    font-weight: bold;
    text-align: right;
    background-color: #ccccff;
  }
  table { 
    border-spacing: 0px;
    border-collapse: collapse;
  }
--->
</STYLE>





# R steps
```{r eval=FALSE, include=FALSE}
#source("https://bioconductor.org/biocLite.R")
#BiocInstaller::biocLite("biocUpgrade")
#BiocManager::install("GO.db")

library("topGO")
library(dplyr)
library(plyr)
library(tidyr)
```

### Read large input table
```{r}
all<-read.csv("/Users/lukas/sciebo/inquilineGenomics18/SOS/data/RELAX2.tsv",sep="\t")
```

### create topGO element for Cluster GO annotations
```{r}

scoID2go<-list()
  for(i in 1:length(all$OGid)){
  scoID2go[[i]]<-unlist(strsplit(as.character(all$GO[i]),", "))
  names(scoID2go)[i]<-as.character(all$OGid[i])
  }

```

### Prepare Objects for GO enrichment analysis
```{r}

scoNames.all<-names(scoID2go)
#restrict gene space to all SCOs orthologs
scoNames<-scoNames.all
scoNames <- factor(scoNames)
names(scoNames) <- scoNames

```



### Summarize things
```{r}
# k<0 & p<0.05
length(unlist(subset(all,p.adjust(pval,method = "fdr")<0.05 & k<1,OGid)))
# k<0 & 0.05< p <0.1
length(unlist(subset(all,p.adjust(pval,method = "fdr")<0.1 & p.adjust(pval,method = "fdr")>0.05 & k<1,OGid)))

# k<0 & p<0.05
length(unlist(subset(all,p.adjust(pval,method = "fdr")<0.05 & k>1,OGid)))
# k<0 & 0.05< p <0.1
length(unlist(subset(all,p.adjust(pval,method = "fdr")<0.1 & p.adjust(pval,method = "fdr")>0.05 & k>1,OGid)))

```


### Select Clusters of Interest from dataframe "all".
```{r}
# create a 0/1 matrix (interesting/not interesting)
######################################

## all parasites
#SCOOfInterest<-unlist(subset(all,p.adjust(P1pv,method = "fdr")<0.05 | p.adjust(AINSpv,method = "fdr")<0.05 | p.adjust(PARGpv,method = "fdr")<0.05 | p.adjust(ACHApv,method = "fdr")<0.05,OGid))

# Origin of parasitism
#SCOOfInterest<-unlist(subset(all,p.adjust(pval,method = "fdr")<0.1 & k<1,OGid))
SCOOfInterest<-unlist(subset(all,p.adjust(pval,method = "fdr")<0.1 & k>1,OGid))

#SCOOfInterest<-unlist(subset(all,p.adjust(PARGpv,method = "fdr")<0.05 ,OGid))
#SCOOfInterest<-unlist(subset(all,p.adjust(ACHApv,method = "fdr")<0.05 ,OGid))


scoList <- factor(as.integer(scoNames %in% SCOOfInterest))

table(scoList)
names(scoList) <- scoNames

relaxedSelSet<-subset(all,OGid %in% SCOOfInterest)

```


```{r}
# create object for Biological Process, Molecular Function and Cellular component:
##################################################################################

GOdata_MF <- new("topGOdata", ontology = "MF", allGenes = scoList, annot = annFUN.gene2GO, gene2GO = scoID2go)
GOdata_BP <- new("topGOdata", ontology = "BP", allGenes = scoList, annot = annFUN.gene2GO, gene2GO = scoID2go)
GOdata_CC <- new("topGOdata", ontology = "CC", allGenes = scoList, annot = annFUN.gene2GO, gene2GO = scoID2go)

# Molecular Function
####################
resultFis_MF 	<- runTest(GOdata_MF, algorithm = "parentchild", statistic = "fisher")
table_MF <- GenTable(GOdata_MF, parentChild = resultFis_MF, orderBy = "parentChild", ranksOf = "parentChild", topNodes = 200)

# Biological Process
####################
resultFis_BP <- runTest(GOdata_BP, algorithm = "parentchild", statistic = "fisher")
table_BP <- GenTable(GOdata_BP, parentChild = resultFis_BP, orderBy = "parentChild", ranksOf = "parentChild", topNodes = 200)


# Cellular Component
####################
resultFis_CC <- runTest(GOdata_CC, algorithm = "parentchild", statistic = "fisher")
table_CC <- GenTable(GOdata_CC, parentChild = resultFis_CC, orderBy = "parentChild", ranksOf = "parentChild", topNodes = 200)


# filter out significant GO terms:
###################################

MF <- subset(table_MF, parentChild < 0.1)
BP <- subset(table_BP, parentChild < 0.1)
CC <- subset(table_CC, parentChild < 0.1)

MF$ontology<-"MF"
BP$ontology<-"BP"
CC$ontology<-"CC"

MF
BP
CC
```
```{r}

subset(all,OGid %in% genesInTerm(GOdata_BP, "GO:0046903")[[1]] & pval<0.05)
subset(all,OGid %in% genesInTerm(GOdata_MF, "GO:0005158")[[1]] & pval<0.05)

```

### save results to file
```{r}
# Origins of social parasitism
res<-rbind(MF,BP) %>% rbind(.,CC)
#outfile<-"/Users/lukas/sciebo/inquilineGenomics18/GOenrichment/results/GOenrichment.RELAX2.parasiteOrigins.relaxedSelection.tsv"
outfile<-"/Users/lukas/sciebo/inquilineGenomics18/GOenrichment/results/GOenrichment.RELAX2.parasiteOrigins.intensifiedSelection.tsv"
write.table(res,outfile,quote=F,sep="\t",row.names = F)
cat("# GO enrichment of genes under relaxed selection (p<0.1, k<1) at any social parasite branch (at FDR<0.1), compared to Ahey, Aech, A1 and A2. \n",file=outfile)
#cat("# GO enrichment of genes under relaxed selection (p<0.05, k<1) at any social parasite branch (at FDR<0.1), compared to all other branches. \n",file=outfile)
write.table(res, outfile,append=TRUE, quote=F,sep="\t",row.names=F)


```


